File: course_3/module_1/examples/job_scheduler.py
class Job:
    """
    Represents a job with a unique ID, weight, and length.
    Attributes:
        job_id (int): The unique identifier for the job.
        weight (float): The weight or importance of the job.
        length (float): The duration required to complete the job.
    """
    def __init__(self, job_id, weight, length):
        if length <= 0:
            raise ValueError("Job length must be positive.")
        if weight < 0:
            raise ValueError("Job weight cannot be negative.")
        self.job_id = job_id
        self.weight = weight
        self.length = length

    @property
    def weight_to_length_ratio(self):
        """
        Calculates the weight-to-length ratio of the job.
        Returns:
            float: The weight-to-length ratio (weight / length).
        """
        return self.weight / self.length

    def __str__(self):
        """
        String representation of the job.
        Returns:
            str: The job details in a readable format.
        """
        return f"Job(id={self.job_id}, weight={self.weight}, length={self.length})"


class JobScheduler:
    """
    Manages the scheduling of jobs using a greedy algorithm.
    Attributes:
        jobs (list[Job]): A list of jobs to be scheduled.
    """
    def __init__(self, jobs):
        """
        Initializes the scheduler with a list of jobs.
        Args:
            jobs (list[Job]): A list of Job objects.
        """
        self.jobs = list(jobs)

    def schedule_jobs(self):
        """
        Schedules jobs based on their weight-to-length ratio in descending order.
        Returns:
            list[Job]: A list of jobs in the optimal execution order.
        """
        self.jobs.sort(key=lambda job: job.weight_to_length_ratio, reverse=True)
        return self.jobs

    def calculate_weighted_completion_time(self):
        """
        Calculates the weighted sum of completion times for the scheduled jobs.
        Returns:
            float: The weighted sum of completion times.
        """
        total_completion_time = 0
        current_time = 0

        for job in self.jobs:
            current_time += job.length  # Add job length to the current time.
            total_completion_time += job.weight * current_time  # Add weighted completion time.

        return total_completion_time


# Example usage.
if __name__ == "__main__":
    # Define a list of jobs with job_id, weight, and length.
    jobs = [
        Job(job_id=1, weight=3, length=2),
        Job(job_id=2, weight=1, length=3),
        Job(job_id=3, weight=2, length=1)
    ]

    # Initialize the job scheduler.
    scheduler = JobScheduler(jobs)

    # Schedule the jobs.
    scheduled_jobs = scheduler.schedule_jobs()
    print("Scheduled jobs (in execution order):")
    for job in scheduled_jobs:
        print(job)

    # Calculate the weighted sum of completion times.
    total_weighted_completion_time = scheduler.calculate_weighted_completion_time()
    print(f"\nWeighted sum of completion times: {total_weighted_completion_time}")

File: course_3/module_1/examples/job_scheduler.go
package main

import (
	"fmt"
	"sort"
)

// Job represents a task with a weight and a length.
type Job struct {
	ID                  int
	Weight, Length      float64
	WeightToLengthRatio float64 // Pre-computed for optimization
}

// NewJob creates a new Job with validation.
func NewJob(id int, weight, length float64) (Job, error) {
	if weight < 0 {
		return Job{}, fmt.Errorf("weight must be non-negative")
	}
	if length <= 0 {
		return Job{}, fmt.Errorf("length must be greater than zero")
	}
	return Job{
		ID:                  id,
		Weight:              weight,
		Length:              length,
		WeightToLengthRatio: weight / length,
	}, nil
}

// Scheduler manages a list of jobs and schedules them using a greedy algorithm.
type Scheduler struct {
	jobs []Job
}

// NewScheduler creates a new Scheduler with a list of jobs.
func NewScheduler(jobs []Job) (*Scheduler, error) {
	if len(jobs) == 0 {
		return nil, fmt.Errorf("job list cannot be empty")
	}
	return &Scheduler{jobs: jobs}, nil
}

// ScheduleJobs sorts the jobs in descending order of their weight-to-length ratio.
func (s *Scheduler) ScheduleJobs() []Job {
	// Create a copy of jobs to avoid modifying the original slice.
	scheduledJobs := make([]Job, len(s.jobs))
	copy(scheduledJobs, s.jobs)

	// Sort jobs by pre-computed weight-to-length ratio in descending order.
	sort.SliceStable(scheduledJobs, func(i, j int) bool {
		return scheduledJobs[i].WeightToLengthRatio > scheduledJobs[j].WeightToLengthRatio
	})
	return scheduledJobs
}

// CalculateWeightedCompletionTime calculates the total weighted sum of completion times.
func (s *Scheduler) CalculateWeightedCompletionTime(jobs []Job) float64 {
	totalCompletionTime := 0.0
	currentTime := 0.0

	for _, job := range jobs {
		currentTime += job.Length
		totalCompletionTime += job.Weight * currentTime
	}
	return totalCompletionTime
}

func main() {
	// Define a list of jobs.
	jobs := []Job{
		{ID: 1, Weight: 3, Length: 2, WeightToLengthRatio: 3.0 / 2.0},
		{ID: 2, Weight: 1, Length: 3, WeightToLengthRatio: 1.0 / 3.0},
		{ID: 3, Weight: 2, Length: 1, WeightToLengthRatio: 2.0 / 1.0},
	}

	// Create a Scheduler instance.
	scheduler, err := NewScheduler(jobs)
	if err != nil {
		fmt.Println("Error creating scheduler:", err)
		return
	}

	// Schedule jobs using the greedy algorithm.
	scheduledJobs := scheduler.ScheduleJobs()

	// Print the scheduled jobs.
	fmt.Println("Scheduled jobs (in execution order):")
	for _, job := range scheduledJobs {
		fmt.Printf("Job(ID=%d, Weight=%.2f, Length=%.2f)\n", job.ID, job.Weight, job.Length)
	}

	// Calculate the total weighted sum of completion times.
	totalWeightedCompletionTime := scheduler.CalculateWeightedCompletionTime(scheduledJobs)
	fmt.Printf("\nWeighted sum of completion times: %.2f\n", totalWeightedCompletionTime)
}
